nexus/auth/src/storage.rs

```diff
@@ -0,0 +1,27 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Describes the dependency from the auth system on the datastore.
+//!
+//! Auth and storage are intertwined, but by isolating the interface from
+//! auth on the database, we can avoid a circular dependency.
+
+use crate::context::OpContext;
+use nexus_db_model::IdentityType;
+use nexus_db_model::RoleAssignment;
+use omicron_common::api::external::Error;
+use omicron_common::api::external::ResourceType;
+use uuid::Uuid;
+
+#[async_trait::async_trait]
+pub trait Storage: Send + Sync {
```

Comment:
I'd be interested in figuring out a way to make this not use async-trait.

I basically went down this route because:

- authz::Context contains one of these objects
- If it uses generics, then `OpContext` also needs a generic parameter
- That changes a *lot* of code
- ... so I went down the `dyn Storage` route instead, as the DataStore is already wrapped in an Arc